src/components/smartSearch/filters/PersonData/index.tsx

```diff
@@ -0,0 +1,195 @@
+    onSubmit: (filter: SmartSearchFilterWithId<PersonDataFilterConfig> | ZetkinSmartSearchFilter<PersonDataFilterConfig>) => void;
+    onCancel: () => void;
+}
+
+interface Criterion {
+    field: DATA_FIELD;
+    value: string;
+}
+
+const PersonData = ({ onSubmit, onCancel, filter: initialFilter }: PersonDataProps): JSX.Element => {
+    const { filter, setConfig, setOp } = useSmartSearchFilter<PersonDataFilterConfig>(initialFilter, { fields: { first_name: '' } });
+
+    const ALL_FIELDS = Object.values(DATA_FIELD);
+
+    const initialCriteria = ALL_FIELDS
+        .filter(f => f in filter.config.fields)
+        .map(f => ({ field: f, value: filter.config.fields[f] || '' }));
+
+    const [criteria, setCriteria] = useState<Criterion[]>(initialCriteria);
+
+    const isSubmittable = criteria.every(c => c.value.length);
+
+    useEffect(() => {
+        const fields = criteria.reduce((acc, c) => {
+            return { ...acc, [c.field]: c.value };
+        }, {});
+        setConfig({
+            ...filter.config,
+            fields,
+        });
+    }, [criteria]);
+
+    const addCriteria = (field: string) => {
+        if (field !== DEFAULT_VALUE) {
+            setCriteria([...criteria, { field: field as DATA_FIELD, value: '' }]);
+        }
+    };
+
+    // event handlers
+    const handleAddFilter = (e: FormEvent) => {
+        e.preventDefault();
+        onSubmit(filter);
+    };
+
+    const handleValueChange = (value: string, c: Criterion) => {
+        setCriteria(criteria.map(criterion => {
+            return criterion.field === c.field ?
+                { ...criterion, value } : criterion;
+        }));
+    };
+
+    const handleSelectChange = (field: string, c: Criterion) => {
+        if (field === DEFAULT_VALUE) {
+            setCriteria(criteria.filter(criterion => criterion.field !== c.field));
+        }
+        else {
+            setCriteria(criteria.map((criterion => {
+                return criterion.field === c.field ?
+                    { ...criterion, field: field as DATA_FIELD } : criterion;
+            })));
+        }
+    };
+
+    // message components
+    const addRemoveSelect = (
+        <StyledSelect onChange={ e => setOp(e.target.value as OPERATION) }
+            value={ filter.op }>
+            <MenuItem key={ OPERATION.ADD } value={ OPERATION.ADD }>
+                <Msg id="misc.smartSearch.person_data.addRemoveSelect.add"/>
+            </MenuItem>
+            <MenuItem key={ OPERATION.SUB } value={ OPERATION.SUB }>
+                <Msg id="misc.smartSearch.person_data.addRemoveSelect.sub" />
+            </MenuItem>
+        </StyledSelect>
+    );
+
+    const getCriteriaString = () => {
+        let existing;
+        let criteriaString: JSX.Element | null = null;
+        criteria.forEach(c => {
+            const options = ALL_FIELDS.filter(f => f === c.field || !criteria.map(c => c.field).includes(f));
```

Comment:
Performance is not really a problem for us here, but code like this still itches a little IMO. This could be optimized in many ways.

Right now there are four loops happening on this line.

1. The `criteria.forEach()` outer loop in which this code is being executed
2. `ALL_FIELDS.filter()` is just a loop that executes a function every iteration
3. The function executes (not always, but almost always) `criteria.map()` which will loop over the entire `criteria` array
4. Finally, after 3 finishes, `includes(f)` will loop for <= N iterations until `f` is found or not found.

I started writing an example of what is actually happening here, but if it's not entirely clear how some of this looping is redundant, lets instead look at it in a debugger. Let me know if you want to do that together!

There are a number of ways that this could be optimized (some of which can be combined, others are incompatible):

* Break `criteria.map()` out of the `criteria.forEach()` loop so that it's only executed once.
* Use `criteria.find()` or `criteria.some()` instead of `criteria.map().includes()`
* Prepare a single list of globally available options that match the `!criteria.map(…).includes(…)` condition, i.e. the ones that are not used anywhere, and use it everywhere instead of creating a new one every time. Append `c.field` (the only that differ) on every iteration.

Reply:
I've optimised it such that the currently selected option will render as the first option, then the global options below it ('remove this field' is always first). It doesn't feel weird to me but if it comes up as strange to use later on, I will change it, have made a note in the overall filter to do list!